nixos/modules/services/development/jupyterhub/default.nix

```diff
@@ -0,0 +1,108 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.services.jupyterhub;
+
+  jupyterhubConfig = pkgs.writeText "jupyterhub_config.py" ''
+    c.JupyterHub.bind_url = "http://${cfg.host}:${toString cfg.port}"
+
+    c.JupyterHub.authentication_class = "${cfg.authentication}"
+    c.JupyterHub.spawner_class = "${cfg.spawner}"
+
+    ${cfg.extraConfig}
+  '';
+
+  jupyterhub_env = pkgs.python3.withPackages (p: with p; cfg.packages);
+in {
+  meta.maintainers = with maintainers; [ costrouc ];
+
+  options.services.jupyterhub = {
+    enable = mkEnableOption "Jupyterhub development server";
+
+    authentication = mkOption {
+      type = types.str;
+      default = "jupyterhub.auth.PAMAuthenticator";
+      description = ''
+        Jupyterhub authentication to use
+      '';
+    };
+
+    spawner = mkOption {
+      type = types.str;
+      default = "systemdspawner.SystemdSpawner";
+      description = ''
+        Jupyterhub spawner to use
+      '';
+    };
+
+    extraConfig = mkOption {
+      type = types.str;
+      default = "";
+      description = ''
+        Extra contents appended to the jupyterhub configuration
+      '';
+    };
+
+    packages = mkOption {
+      type = types.listOf types.package;
+      default = [
+        pkgs.python3Packages.jupyterhub
+        pkgs.python3Packages.jupyterhub-systemdspawner
+      ];
+      description = ''
+        Python environment to run jupyterhub
+      '';
+    };
+
+    # TODO do work here
```

Comment:
This is the part that can be shared with the `jupyter` module.

Reply:
Yes this work now shares kernel instantiation with the jupyter module.